### PR TITLE
Internals documentation for the internal service provider

### PR DESCRIPTION
**Pull request by ajcvickers**

### CODE REVIEW

**Comment by ajcvickers** [comment]
@aspnet/efteam This is the start of work on internals/architecture documentation. I'm doing this in a feature branch because I'm not sure what the overall structure will look like until I have more content. All links are currently just placeholders for where I want to link to something.

File: entity-framework/core/internals/internal-service-provider.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+title: The internal service provider - EF Core
```
Comment (roji) [inline entity-framework/core/internals/internal-service-provider.md]: Nit: I think we capitalize our top-level titles

File: entity-framework/core/internals/internal-service-provider.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+author: ajcvickers
+ms.date: 01/25/2020
+uid: core/internals/internal-service-provider
+---
+
+# The internal service provider
```
Comment (roji) [inline entity-framework/core/internals/internal-service-provider.md]: Same nit re capitalization

File: entity-framework/core/internals/internal-service-provider.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+The application must then handle caching.
+
+### Scopes and lifetimes
+
+A new [service provider scope]() is created for each DbContext instance.
+This means that:
+* [Singleton]() services are shared between DbContext instances
+* [Scoped]() services are shared within, but not between, DbContext instances
+
+A DbContext instance typically represents a [session]() or [unit-of-work]().
```
Comment (roji) [inline entity-framework/core/internals/internal-service-provider.md]: Maybe say a word about DI scopes at the beginning of this section before going into DbContext etc.?

File: entity-framework/core/internals/internal-service-provider.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+Most singleton services behave the same way between different context instances.
+That is, changes in the context configuration do not affect these services.
+
+However, some changes in context configuration require singleton services to behave differently.
+For example, [sensitive data logging]() requires singleton services to change the way they log.
+Likewise, some changes in context configuration require different service implementations.
+For example, [changing the database provider]() requires different implementations for many services.
+
+These additional service providers are cached automatically.
```
Comment (roji) [inline entity-framework/core/internals/internal-service-provider.md]: Go into ISingletonOptions here or is it too much? Am worried this is a bit too abstract for newbies.

File: entity-framework/core/internals/internal-service-provider.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+Both these resolutions must return _the same service instance_.
+This is achieved redirecting resolution for ITypeMappingSource to use IRelationalTypeMappingSource:
+
+```CSharp
+TryAdd<ITypeMappingSource>(p => p.GetService<IRelationalTypeMappingSource>());
+```
+
+## Resolving services
+
+Service implementations can be public or [public but internal]().
```
Comment (roji) [inline entity-framework/core/internals/internal-service-provider.md]: Maybe introduce the idea of pubternal before, including some of the reasoning (in a more general intro not specific to the internal service provider?)

File: entity-framework/core/internals/internal-service-provider.md (hunk 3 above)
Comment (ajcvickers) [reply entity-framework/core/internals/internal-service-provider.md]: What I'm hoping here is that I can find links to these concepts in the D.I. documentation. I decided not to try to describe D.I. in general here. Depending what I find to link to I will revisit this.

File: entity-framework/core/internals/internal-service-provider.md (hunk 4 above)
Comment (ajcvickers) [reply entity-framework/core/internals/internal-service-provider.md]: I think that should go into a different page--I created a list of possible things to cover here: https://github.com/aspnet/EntityFramework.Docs/issues/1920

File: entity-framework/core/internals/internal-service-provider.md (hunk 5 above)
Comment (ajcvickers) [reply entity-framework/core/internals/internal-service-provider.md]: The the dummy link here will ultimately go to a page on that.

**Comment by ajcvickers** [comment]
Closing for now; will bring it back in a new format later.

**Pull request closed without merging**